externs/browser/url.js

```diff
@@ -22,10 +22,20 @@
  * @author rdcronin@google.com (Devlin Cronin)
  */
 
+/**
+ * @typedef {IArrayLike<string>}
```

Comment:
@brad4d - is there a way to tell the compiler that this array/iterable must have a length of 2? Chrome throws an exception otherwise, `Sequence initializer must only contain pair elements`.
I think there will be other cases where the length matters. For example, I assume the upcoming Object.fromEntries would require it.

And yes, it must be an `Iterable`.

Reply:
Sadly there's no way to say "must be array/iterable of length 2".

Someday maybe we'll finally have time to implement tuple types. *(wistful sigh)*

FYI, I'm a bit concerned that the spec is defined purely in terms of `sequence`, when I couldn't find a hard definition of what that translates to in JS terms.

I'm inclined to go with this PR as it currently exists, but I do think we have this possible failure case.

1. code input to the compiler uses a non-Array Iterable object for the individual tuples or the list of tuples.
2. after compilation this code runs on an older browser that assumed it would be getting an Array, so it fails, because it doesn't know to use the `[Symbol.iterator]` property for iteration and just tries to use a counting loop with a `length` property.

This doesn't seem terribly likely, since I expect people will just always use Arrays.
That's what's most convenient, after all.

Reply:
Seems pretty defined to me -
https://heycam.github.io/webidl/#es-sequence

<strike>Also, I believe `URLSearchParams` was implemented by the time browsers already supported iterables (and even more so with this constructor overload that expects a sequence as a parameter), so it is not possible non-array iterables are not supported. Either sequences are not supported at all, or they are fully supported.</strike>
However... I just tested Edge 17 and 18 (going down soon, though. And 16 did not support `URLSearchParams` at all) - Edge does not support iterables, only arrays -
```js
new URLSearchParams((a=new Map(),a.set('a', '1'),a)).toString()
// "" in Edge
// "a=1" in Chrome 71, Firefox 64 and Safari 12
new URLSearchParams([(a=new Set(),a.add('a'),a.add('b'),a)]).toString()
// Invalid argument in Edge
// "a=b" in Chrome 71, Firefox 64 and Safari 12
```

Should I change it to array of arrays to cater for poor, poor Edge?

Reply:
No, I don't think we should change it to an array of arrays.

As you said, Edge 16 doesn't support URLSearchParams at all.

We don't polyfill web-standard stuff, just core JS features, so it is expected that when you write code using web-standard stuff you still have to be concerned about what your customers' browsers actually support. Sadly we just can't do everything. :)

Reply:
@brad4d - I am not sure how the fact that Edge 16 does not support this at all affects the decision. Edge 17 and 18 (but not the future Chromium based Edge, obviously) do not support Iterable in this case at all which is exactly what you were worried about before in https://github.com/google/closure-compiler/pull/3033#discussion_r251611405. Edge 17 and 18 will be here for a while (with a low market share, but still).

(I have not tested old versions of other browsers, so old versions may have had Array and not Iterable support in this case, too, at some point)

I am not necessarily advocating Array and not Iterable here (I do have a slight preference for lessening the footguns by supporting only Array), I am just verifying that you reconsidered your previous worries in light of the new information about Edge turning out supporting exactly what you were worried about.

Reply:
@brad4d - also, this is not a polyfill. The externs do try to reflect the lowest common denominator in many cases in order to prevent footguns by naïve developers.

Reply:
@phistuck You've convinced me.

I think the small amount of pain felt by a few developers who really want to use Iterables here is more than balanced by the very large amount of potential pain caused by people using Iterables and having their code mysteriously fail on Edge.

Of course, their Array-using code will mysteriously fail on even earlier versions of Edge already, because those don't support `URLSearchParam` at all, but, why make things even more dangerous?

I will await your update for import.
Thanks.